eos-upgrade-eos2-to-eos3

```diff
@@ -0,0 +1,222 @@
+    if [ ${major_version} == 2 ]; then
+	echo "Please update the OS and try again"
+    elif [ ${major_version} == 3 ]; then
+	echo "You already are running Endless OS version 3"
+    else
+	echo "You are running an unsupported version: ${version}"
+    fi
+    exit 1
+fi
+
+if ! $FORCE; then
+    echo
+    echo "You are about to upgrade Endless OS from version 2 to version 3."
+    echo
+    echo "Endless OS 3 uses a new technology for applications called flatpak,"
+    echo "which is a standard format for deploying Linux applications that enables"
+    echo "Endless to run applications distributed by other development teams."
+    echo
+    echo "The legacy application format is no longer supported, and the first step"
+    echo "of the upgrade process is to remove all installed application bundles."
+    echo "Then, roughly a gigabyte of data will be downloaded from the internet,"
+    echo "which may take a long time and incur data charges depending on your"
+    echo "internet connection and contract with your provider."
+    echo
+    echo "In the meantime, while downloading Endless OS 3, you may continue"
+    echo "to use Endless OS 2 and some of the core applications such as"
+    echo "the internet browser and office applications."
+    echo "Once the upgrade to Endless OS 3 is downloaded and ready, you will be"
+    echo "asked to reboot, and then you will be able to install new flatpak apps"
+    echo "from the Endless OS 3 app center."
+    echo
+    echo "Once the download completes, it will take several additional minutes"
+    echo "for the deployment to complete -- please be patient and wait until"
+    echo "you see the 'Upgrade complete!' message."
+    echo
+    read -p "Are you sure you want to remove all legacy app bundles and proceed? [y/N] "
+    response=${REPLY,,} # to lower
+    if [[ ! $response =~ ^(yes|y)$ ]]; then
+        exit 1
+    fi
+fi
+
+# Stop any updates in progress
+killall ostree-daemon || true
+killall eos-updater || true
+
+# For older installs (eos2.3.x and earlier), we had a separate boot
+# partition which we no longer support on eos3
+if [ ! "$(ls -A /sysroot/boot)" ]; then
+    echo "Migrating separate boot partition into main partition"
+    # Get the boot partition (e.g., /dev/sda1 or /dev/mmcblk0p1)
+    boot_part=`mount | grep /boot | awk '{print $1}'`
+    # Extract the device name (e.g., /dev/sda or /dev/mmblk0)
+    boot_dev=`echo $boot_part | sed 's/p*[0-9]\+$//'`
+    # Copy the contents from /boot to the main partition
+    cp -a /boot/* /sysroot/boot
+    # Manually unmount the boot partition
+    umount /boot
+    # Re-install the bootloader, telling it to look for config files
+    # in the main partition
+    grub-install --boot-directory=/sysroot/boot --modules="ext2 part_msdos part_gpt" --directory=/usr/lib/grub/i386-pc $boot_dev
+    sync
+    # Remove the old boot partition
+    sfdisk --dump $boot_dev | grep -v $boot_part | sfdisk --force $boot_dev
```

Comment:
@dsd: Oops -- sorry about that.  I suppose this will be more helpful...

```
endless@endless:~/Downloads$ sudo ./strace sfdisk --dump /dev/mmcblk0
execve("/usr/sbin/sfdisk", ["sfdisk", "--dump", "/dev/mmcblk0"], [/* 16 vars */]) = 0
brk(0)                                  = 0x9ae2000
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
mmap2(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0xf76f8000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat64(3, {st_mode=S_IFREG|0644, st_size=138078, ...}) = 0
mmap2(NULL, 138078, PROT_READ, MAP_PRIVATE, 3, 0) = 0xf76d6000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/i386-linux-gnu/libfdisk.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0 f\0\0004\0\0\0"..., 512) = 512
fstat64(3, {st_mode=S_IFREG|0644, st_size=391684, ...}) = 0
mmap2(NULL, 394644, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0xf7675000
mmap2(0xf76d2000, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x5d) = 0xf76d2000
mmap2(0xf76d5000, 1428, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0xf76d5000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/i386-linux-gnu/libsmartcols.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0P:\0\0004\0\0\0"..., 512) = 512
fstat64(3, {st_mode=S_IFREG|0644, st_size=157652, ...}) = 0
mmap2(NULL, 160596, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0xf764d000
mmap2(0xf7671000, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x24) = 0xf7671000
mmap2(0xf7674000, 852, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0xf7674000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/i386-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\1\1\1\3\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0\340\233\1\0004\0\0\0"..., 512) = 512
fstat64(3, {st_mode=S_IFREG|0755, st_size=1753800, ...}) = 0
mmap2(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0xf764c000
mmap2(NULL, 1755772, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0xf749f000
mmap2(0xf7646000, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1a7) = 0xf7646000
mmap2(0xf7649000, 10876, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0xf7649000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/i386-linux-gnu/libuuid.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0p\17\0\0004\0\0\0"..., 512) = 512
fstat64(3, {st_mode=S_IFREG|0644, st_size=17984, ...}) = 0
mmap2(NULL, 20680, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0xf7499000
mmap2(0xf749d000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3) = 0xf749d000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/i386-linux-gnu/libblkid.so.1", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\1\1\1\0\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0\240L\0\0004\0\0\0"..., 512) = 512
fstat64(3, {st_mode=S_IFREG|0644, st_size=297420, ...}) = 0
mmap2(NULL, 300380, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0xf744f000
mmap2(0xf7495000, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x46) = 0xf7495000
mmap2(0xf7498000, 1372, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0xf7498000
close(3)                                = 0
mmap2(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0xf744e000
mmap2(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0xf744d000
set_thread_area({entry_number:-1 -> 12, base_addr:0xf744d740, limit:1048575, seg_32bit:1, contents:0, read_exec_only:0, limit_in_pages:1, seg_not_present:0, useable:1}) = 0
mprotect(0xf7646000, 8192, PROT_READ)   = 0
mprotect(0xf749d000, 4096, PROT_READ)   = 0
mprotect(0xf7495000, 8192, PROT_READ)   = 0
mprotect(0xf7671000, 8192, PROT_READ)   = 0
mprotect(0xf76d2000, 4096, PROT_READ)   = 0
mprotect(0x805e000, 8192, PROT_READ)    = 0
mprotect(0xf771d000, 4096, PROT_READ)   = 0
munmap(0xf76d6000, 138078)              = 0
ioctl(0, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
brk(0)                                  = 0x9ae2000
brk(0x9b03000)                          = 0x9b03000
open("/usr/lib/locale/locale-archive", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3
fstat64(3, {st_mode=S_IFREG|0644, st_size=114327312, ...}) = 0
mmap2(NULL, 2097152, PROT_READ, MAP_PRIVATE, 3, 0) = 0xf724d000
mmap2(NULL, 1249280, PROT_READ, MAP_PRIVATE, 3, 0x163) = 0xf711c000
mmap2(NULL, 4096, PROT_READ, MAP_PRIVATE, 3, 0x1256) = 0xf76f7000
close(3)                                = 0
ioctl(1, SNDCTL_TMR_TIMEBASE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
openat(AT_FDCWD, "/sysroot/home/endless/.config/terminal-colors.d", O_RDONLY|O_NONBLOCK|O_LARGEFILE|O_DIRECTORY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/terminal-colors.d", O_RDONLY|O_NONBLOCK|O_LARGEFILE|O_DIRECTORY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/dev/mmcblk0", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3
fadvise64_64(3, 0, 0, POSIX_FADV_RANDOM) = 0
fstat64(3, {st_mode=S_IFBLK|0660, st_rdev=makedev(179, 0), ...}) = 0
uname({sys="Linux", node="endless", ...}) = 0
ioctl(3, BLKGETSIZE64, 0x9ae5254)       = 0
open("/sys/dev/block/179:0", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 4
openat(4, "dm/uuid", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
close(4)                                = 0
open("/sys/dev/block/179:0", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 4
fstatat64(4, "partition", 0xffa059dc, 0) = -1 ENOENT (No such file or directory)
openat(4, "dm/uuid", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
close(4)                                = 0
ioctl(3, CDROM_GET_CAPABILITY, 0)       = -1 EINVAL (Invalid argument)
ioctl(3, BLKALIGNOFF, 0xffa06b88)       = 0
ioctl(3, BLKIOMIN, 0xffa06b88)          = 0
ioctl(3, BLKIOOPT, 0xffa06b88)          = 0
ioctl(3, BLKPBSZGET, 0xffa06b88)        = 0
ioctl(3, BLKSSZGET, 0x9ae526c)          = 0
ioctl(3, BLKSSZGET, 0xffa06c68)         = 0
uname({sys="Linux", node="endless", ...}) = 0
ioctl(3, BLKGETSIZE64, 0xffa06c20)      = 0
ioctl(3, 0x301, 0xffa06c24)             = 0
_llseek(3, 0, [0], SEEK_SET)            = 0
read(3, "\353c\220\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 512) = 512
fstat64(1, {st_mode=S_IFCHR|0620, st_rdev=makedev(136, 0), ...}) = 0
mmap2(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0xf76f6000
write(1, "label: dos\n", 11label: dos
)            = 11
write(1, "label-id: 0x00000000\n", 21label-id: 0x00000000
)  = 21
write(1, "device: /dev/mmcblk0\n", 21device: /dev/mmcblk0
)  = 21
write(1, "unit: sectors\n", 14unit: sectors
)         = 14
write(1, "\n", 1
)                       = 1
write(1, "/dev/mmcblk0p1 : start=        6"..., 74/dev/mmcblk0p1 : start=        6144, size=      188416, type=83, bootable
) = 74
write(1, "/dev/mmcblk0p2 : start=      194"..., 64/dev/mmcblk0p2 : start=      194560, size=    30590976, type=83
) = 64
close(3)                                = 0
close(1)                                = 0
munmap(0xf76f6000, 4096)                = 0
close(2)                                = 0
exit_group(0)                           = ?
```